### DISCUSSION TITLE
Setting Up VSCode Development Environment For MonoGame

### DISCUSSION THREAD

**Discussion by AristurtleDev** [MonoGame/MonoGame#8131] [Desktop computers]
# Setting Up VSCode Development Environment For MonoGame
This is intended to be a quick reference guide on setting up your development environment to create games using MonoGame with VSCode.

> 💡**Note**
> This guide will assume that you have not setup .NET or VSCode prior to reading

# Video
The following video reference is provided which goes over the information covered here, but the video is Windows specific.  You can follow along on Mac and/or Linux, just be aware there may be OS difference not explained in the video, but they are covered in the guide below.

[![MonoGame and VSCode Development Environment Setup](https://img.youtube.com/vi/WKfufhjd_5o/0.jpg)](https://www.youtube.com/watch?v=WKfufhjd_5o)
https://www.youtube.com/watch?v=WKfufhjd_5o&t=13s

# Table of Contents
1. [Download .NET SDK](#download-sdk)
2. [Install MonoGame `dotnet new` Templates](#install-templates)
3. [Install Visual Studio Code](#install-vscode)
4. [Install C# Extensions in Visual Studio Code](#install-extensions)
5. [Create A New Project](#create-new-project)
6. [Opening the MGCB Editor](#open-mgcb)
7. [Running/Debugging Your Project](#run-debug)
8. [Caveats and Common Gotchas](#gotchas)

# 1. Download .NET SDK <a id="download-sdk" style="display: none;">
In order to develop anything with C#, you will need the .NET SDK installed.  At the time of this writing, the current version is .NET 8.0.  Below is the setup instructions for each operating system, click to expand and see the instructions for your operating system.

<details><summary>Windows Steps</summary>

![Download .NET 8.0 SDK Windows](https://github.com/MonoGame/MonoGame/assets/103014489/0a169f06-b790-4ddd-bc8f-6401d83da67f)

Windows users can download the .NET SDK from the website https://dotnet.microsoft.com/en-us/download

Once you have downloaded and installed, open a command prompt and type the following command to ensure installation is good

</details>

<details><summary>Mac Steps</summary>

![Download .NET 8.0 SDK Mac](https://github.com/MonoGame/MonoGame/assets/103014489/88e2957e-39af-4779-937c-a4dcc416695f)

Mac users can download the .NET SDK from the website https://dotnet.microsoft.com/en-us/download.  Ensure you choose the appropriate version based on the type of mac you have.

> ⚠️Warning
> Do not install the dotnet sdk using `brew`.  Doing this will install it as read-only and require sudo to perform certain actions such as downloading nuget packages.  

> 💡Apple Silicon
> For Apple Silicon (M1/M2) Macs, please note there is currently a [known issue for building content using the MonoGame Content Builder](https://docs.monogame.net/articles/whats_new#apple-silicon-m1-support).  

</details>

<details><summary>Linux Steps</summary>

![Download .NET 8.0 SDK Linux](https://github.com/MonoGame/MonoGame/assets/103014489/320f8204-a34d-4f07-b304-5e45f711f48c)

Linux users will need to perform a manual installation of the .NET SDK.  You can find instructions for various Linux systems here: https://learn.microsoft.com/en-us/dotnet/core/install/linux?WT.mc_id=dotnet-35129-website

Regardless of which version of Linux you are using, you can also use the scripted install, which is the method I would recommend using.  You can find information on doing that at https://learn.microsoft.com/en-us/dotnet/core/install/linux-scripted-manual#scripted-install

> 💡**Note**
> After doing the scripted installation, ensure that you setup your `$PATH` environment variable so that it includes the newly installed dotnet.  Instructions can be found at https://learn.microsoft.com/en-us/dotnet/core/install/linux-scripted-manual#set-environment-variables-system-wide

</details>

Once you have finished installing the .NET SDK for your operating system, open a command prompt/terminal and enter the following command to ensure everything is setup correctly

```sh
dotnet --info
```

If everything is setup correctly, you should get output with information about the installation.

![Successful .NET SDK Installation](https://github.com/MonoGame/MonoGame/assets/103014489/6a5e8c81-7f77-4e75-908f-814dbe480a19)

# 2. Install MonoGame `dotnet new` Templates <a id="install-templates" style="display: none;">
The .NET SDK installation you just performed comes with various project templates by default.  You can view these templates by entering the following in a command prompt/terminal

```sh
dotnet new list
```

![Default dotnet new templates ](https://github.com/MonoGame/MonoGame/assets/103014489/37f9fbbf-a338-42dd-bfb3-858fe50f1516)

In order to add the MonoGame templates to this, you will need to tell dotnet to install them.  To do this, open a command prompt/terminal and enter the following command:

```sh
dotnet new install MonoGame.Templates.CSharp
```
![MonoGame Template Install Command Results](https://github.com/MonoGame/MonoGame/assets/103014489/c77ec950-178b-4474-ac87-213a89c78f19)

> 💡 **Note**
> You will see an error when installing the templates that it failed to load the UWP template.  You can ignore this error, the UWP template is being deprecated in a future version of MonoGame.

This command installs the following templates

<a id="project-type-table" style="display: none;"></a>

**Project Type Table**
| Name | Short Name | Documentation |
|---|---|---|
| MonoGame Android Application | `mgandroid` | https://monogame.net/articles/platforms.html#android |
| MonoGame Content Pipeline Extension | `mgpipeline`  | https://monogame.net/articles/platforms.html#content-pipeline-extension|
| MonoGame Cross-Platform Desktop Application | `mgdesktopgl` | https://monogame.net/articles/platforms.html#desktopgl |
| MonoGame Game Library | `mglib` | https://monogame.net/articles/platforms.html#net-standard-library |
| MonoGame iOS Application | `mgios` | https://monogame.net/articles/platforms.html#ios|
| MonoGame Shared Library Project | `mgshared` | https://monogame.net/articles/platforms.html#shared-project |
| MonoGame Windows Desktop Application | `mgwindowsdx` | https://monogame.net/articles/platforms.html#windowsdx|
| MonoGame Windows Universal XAML Application | `mguwpxaml` | https://monogame.net/articles/platforms.html#windowsuniversal |

The **Short Name** is what will be used when creating a new project.  We'll discuss this in more detail in a moment.

# 3. Install Visual Studio Code <a id="install-vscode" style="display: none;">
Next you will need to install Visual Studio Code.  You can download it for your operating system from https://code.visualstudio.com/Download/

![Download Visual Studio Code](https://github.com/MonoGame/MonoGame/assets/103014489/1267a2e9-063f-42b0-bac1-da0f23ff8566)

> 💡Mac/Linux Users
> After installing Visual Studio code, you may wish to add the `code` command to your `$PATH` environment variable.  Visual Studio Code makes this simple for you to do.
> 1. Open Visual Studio Code
> 2. Open the **Command Palette** by pressing
>     - Mac: `Command + Shift + P`
>     - Linux: `Control + Shift + P`
> 3. In the **Command Palette** type `shell comand`, then select the option **Shell Command: Install 'code' command in PATH**
>
> ![Command_Palette_Add_Path](https://github.com/MonoGame/MonoGame/assets/103014489/7a552074-2a57-478b-89fe-627657ecfeee)

# 4. Install C# Extensions in Visual Studio Code <a id="install-extensions" style="display: none;">
Visual Studio Code by itself is a glorified text editor powered by extensions.  For developing C# applications in Visual Studio Code, you will need to install the C# extensions.  This will give you various benefits such as syntax highlighting, code analysis, intellisense, and easy to use methods to run and debug your game.

To install the extensions first click the **Extensions** icon in the **Activity Bar** on the left

![Click Extensions in Activity Bar](https://github.com/MonoGame/MonoGame/assets/103014489/96a4c9a4-3d7e-496b-b40f-8bef800e9fc9)

Next, in the **Search Box** type `C#`, then select the **C# Dev Kit** extension and install it.  Installing this will install both the **C# Dev Kit** extension as well as the base **C#** extension

![Install C# Dev Kit Extension](https://github.com/MonoGame/MonoGame/assets/103014489/5ddeafb5-2b1f-4dab-8460-6c99c8faac7d)

# 5. Create A New Project <a id="create-new-project" style="display: none;">
Now that we have our development environment setup, the next step is to create a new project.  Creating a new MonoGame project when using Visual Studio Code requires you to be comfortable using a command prompt/terminal.  It's not any crazy commands, just know that's what you'll use to create the new projects.

1. Create a new folder/directory somewhere that will hold your MonoGame game project files.
2. Open Visual Studio Code
9. Choose `File -> Open Folder` and open the folder you just created

If done correctly, you should see the name of your folder open in the **Explorer** panel on the left side of Visual Studio Code.

![Folder Opened in VSCode](https://github.com/MonoGame/MonoGame/assets/103014489/ccafceb7-791f-457f-a0c2-4f153c3ccd5b)

Next, you'll need to open the **Terminal Panel** in Visual Studio Code so we can enter the commands to create a new MonoGame project.  You can open the terminal by selecting `View > Terminal` from the top menu in Visual Studio Code, or by using the `CTRL + ~` keyboard shortcut.  Once opened, it will open at the bottom

![Terminal Panel](https://github.com/MonoGame/MonoGame/assets/103014489/2d2a7c15-f6e1-41bc-b98a-08a9302bb820)

Now that the **Terminal Panel** is open, enter the following commands to setup the typical project structure, including a Solution File (.sln)

```sh
dotnet new sln -n MyGame
dotnet new mgdesktopgl -n MyGame
dotnet sln add MyGame
dotnet restore
```

> 💡 **Recommended**
> In the commands above, replace the `MyGame` string with the name you want to use for your Project (game).

The commands above, in order, are performing the following actions
1. `dotnet new sln -n MyGame` creates a new Solution File (.sln) called MyGame
2. `dotnet new mgdesktopgl -n MyGame` creates the MonoGame project based on the project type given.
3. `dotnet sln add MyGame` this will add the MonoGame project that was just created to the Solution File that was created
10. `dotnet restore` performs a restore on the project so that all NuGets and tools are restored and everything is ready to go.

> 💡 **Choosing Project Type**
> In the commands above, we created a new project using the command `dotnet new mgdesktopgl`.  `mgdesktopgl` will create a new **MonoGame Crossplatform DesktopGL** project.  Refer to the [Project Type Table](#project-type-table) for a the short names to use for different project types.

If all commands were entered correctly, you should now have your project structure setup in the **Explorer Panel** similar to the image below

![Project Setup](https://github.com/MonoGame/MonoGame/assets/103014489/34dfd9a4-14d2-461f-8ec2-6e1aaf4eba15)

# 6. Opening the MGCB Editor <a id="open-mgcb" style="display: none;">
If you are going to use the [MGCB Editor](https://monogame.net/articles/content_pipeline/using_mgcb_editor/) for content management, there are two ways you can open your `Content.mgcb` file.

The first method is to just enter the command in the **Terminal Panel** to open it

```sh
dotnet mgcb-editor ./Content/Content.mgcb
```

When doing this, ensure that the current directory in the terminal is the same directory as the project (.csproj) file.

The other method is to use the **MonoGame Content Builder (Editor)** extension created by mangrimen. I do not use this extension personally so I don't have any feedback to offer on it, however do note this is a third party developed tool and not developed or maintained by the MonoGame Foundation

![MonoGame Content Builder (Editor) Extension](https://github.com/MonoGame/MonoGame/assets/103014489/d4080719-a3ba-4fc0-9403-bc999f7496e9)

# 7. Running/Debugging Your Project <a id="run-debug" style="display: none;">
When you installed the **C# Dev Kit** extension earlier, this adds a new **Solution Explorer** panel in Visual Studio Code whenever you have a folder open that contains a C# Solution file.  This **Solution Explorer** panel works similar to the same one you would have in the full Visual Studio IDE.

If you want to run and/or debug your project, you can just right-click the project in the **Solution Explorer** panel and choose `Debug > Start New Instance`

![Run/Debug Project](https://github.com/MonoGame/MonoGame/assets/103014489/6f0afa22-d946-4e06-8b53-77c8c74fef93)

# 8. Caveats and Common Gotchas <a id="gotchas" style="display: none;">
When using Visual Studio Code to develop games in MonoGame, there are a few things to be aware of when comparing the experience to using something like Visual Studio or Rider.

First, you will need to be somewhat familiar with using the dotnet command line interface (cli) to do things such as running your project (`dotnet run`) or adding third party libraries via NuGet packages (`dotnet add package`).  Understanding the basic concepts of using the dotnet cli will go a long way in helping you when using Visual Studio Code. You can find more information at https://learn.microsoft.com/en-us/dotnet/core/tools/

Secondly, when opening the terminal in Visual Studio Code (`Control + ~`), the terminal panel will open and automatically set the directory path to the path of the root folder that you opened in Visual Studio Code.  Depending on how you have your project structure setup, remember that you need to have the path in the Project (.csproj) directory when using dotnet cli commands that affect the project or when opening the MGCB Editor.  So be sure you cd to the correct directory if needed too.

Finally, at the time of this writing, I do not have a good solution to give you for using Visual Studio Code for mobile game development.  You can absolutely use it to do this, however the things that need to be setup for debugging purposes are beyond the scope of this guide at the moment.  If you are using Windows, I fully recommend to instead use Visual Studio, since it has native support for just plugging your mobile device in and debugging the application live on your device.

**Comment by ThomasFOG** [comment]
"quick guide"... more like "really cool tutorial".

**Reply by Apostolique** [reply]
Pretty cool! More in-depth than mine that's for sure! <https://learn-monogame.github.io/how-to/develop-vscode/>

**Reply by AristurtleDev** [reply]
Yea I keep saying "quick" in my guides/videos and they are never quick.

Thanks

**Comment by AshtonStephens** [comment]
> > 💡Apple Silicon
For Apple Silicon (M1/M2) Macs, please note there is currently a [known issue for building content using the MonoGame Content Builder (mgcb)](https://monogame.net/articles/whats_new/#apple-silicon-m-support).

@AristurtleDev the docs migration broke the above link, it looks like this is the right location now: https://docs.monogame.net/articles/whats_new#apple-silicon-m1-support

**Reply by AristurtleDev** [reply]
Thanks
